src/components/shared/CompactableListWithRemoveButton.js

```diff
@@ -0,0 +1,138 @@
+      result[result.length - 2] = ' and '; // replace last comma with 'and'
+      return result;
+    }
+  }
+}
+
+type Props = {|
+  +className?: string,
+  +compact: boolean,
+  +items: string[],
+  +buttonTitle: string,
+  +showIntroduction: string,
+  +onItemRemove: number => mixed,
+|};
+
+export default class CompactableListWithRemoveButton extends PureComponent {
+  props: Props;
+
+  constructor(props: Props) {
+    super(props);
+
+    (this: any)._onRemoveButtonMouseDown = this._onRemoveButtonMouseDown.bind(
+      this
+    );
+  }
+
+  // We use the mousedown event so that we can avoid that the input loses the focus.
+  _onRemoveButtonMouseDown(
+    e: SyntheticMouseEvent & { currentTarget: HTMLButtonElement }
+  ) {
+    e.preventDefault(); // prevents losing the focus
+    this.props.onItemRemove(+e.currentTarget.dataset.index);
+  }
+
+  _renderFull() {
+    const { items, buttonTitle, showIntroduction } = this.props;
+
+    if (showIntroduction && !items.length) {
+      return (
+        <div className="listWithRemoveButton_introduction">
+          {showIntroduction}
+        </div>
+      );
+    }
+
+    return (
+      <TransitionGroup className="listWithRemoveButton_full">
+        {items.map((item, i) =>
+          <CSSTransition
+            key={item}
+            classNames="listWithRemoveButton_full_transition"
+            timeout={300}
+          >
+            <div className="listWithRemoveButton_full-itemBox">
+              <div className="listWithRemoveButton_full-item">
+                {item}
+              </div>
+              <button
+                className="listWithRemoveButton_full-button"
+                title={buttonTitle}
+                aria-label={buttonTitle}
+                data-index={i}
+                type="button"
+                onMouseDown={this._onRemoveButtonMouseDown}
+              >
+                ✕
+              </button>
+            </div>
+          </CSSTransition>
+        )}
+      </TransitionGroup>
+    );
+  }
+
+  _renderCompact() {
+    const { items } = this.props;
+    if (!items.length) {
+      return null;
+    }
+
+    return (
+      <div className="listWithRemoveButton_compact">
+        <div className="listWithRemoveButton_compact-itemList" key="itemList">
+          {englishListJoin(
+            items.map(item =>
+              <span key={item}>
+                {item}
+              </span>
+            )
+          )}
+        </div>
+        <div className="listWithRemoveButton_compact-itemCount" key="itemCount">
+          ({items.length})
+        </div>
+      </div>
+    );
+  }
+
+  render() {
+    const { className, compact } = this.props;
+
+    const classes = classNames(
+      'listWithRemoveButton',
+      className,
+      `listWithRemoveButton_${compact ? 'isCompact' : 'isFull'}`
+    );
+
+    return (
+      <div className={classes}>
+        {this._renderCompact()}
+        {this._renderFull()}
+      </div>
```

Comment:
overtime I switched between this logic (put everything in the DOM and hide with CSS) or the logic of displaying only what's needed. Now I settled on this one because I thought it would be easier to transition between both -- but in the end, no transition :) So tell me if you prefer the other logic.

My advice is that we can still add the transition later so I'd rather keep this logic.